library(edgeR)
library(org.Hs.eg.db)
library(clusterProfiler)
library(heatmaps)
library(gplots)
library(dplyr)
library(pheatmap)
library(stringr)
library(grid)

setwd("C:/Users/37140/Documents/")
#read the TPM file generated by the rsem quantification process 
TPM<-read.csv("./gene-count-matrix.txt", stringsAsFactors = FALSE,na.strings = "",sep = "",fill=T,row.names = 1)

#transform the data form, for character to numeric 
TPM <- TPM[,]  %>% mutate_if(is.character, ~as.numeric(as.character(.)))

#filter out the lower expression gene
TPM=TPM[rowSums(cpm(TPM) > 1) >= 2,]

#rename the normal patients
row.names(TPM)<-str_sub(rownames(TPM),1,15)
colnames(TPM)[22:23]<-c("NNL.143.genes","NNL.297.genes")

#add the SYMBOL ID
TPM$ENSEMBL<-row.names(TPM)
entrezIDs <- bitr(geneID = rownames(TPM),fromType = "ENSEMBL",toType = "SYMBOL",OrgDb ="org.Hs.eg.db" ,drop = T) 
TPM<-inner_join(TPM,entrezIDs,by="ENSEMBL")

#for those symbols have more than one row, aggregate them and take their mean value
TPM<-aggregate(x=TPM[,1:23],by=list(TPM$SYMBOL),mean,na.rm=TRUE) #2
rownames(TPM)<-TPM$Group.1
TPM<-TPM %>% dplyr::select(-Group.1)

#dividing the patients into three subgroups 
subgroups<-c("BT-BTN","BT-NORM","BTN-NORM")
BT_BTN<-dplyr::select(TPM,starts_with("BT")|starts_with("BTN"))
BT_NNL<-dplyr::select(TPM,starts_with("BT.")|starts_with("NNL"))
BTN_NNL<-dplyr::select(TPM,starts_with("BTN")|contains("NNL"))
dat<-list(BT_NNL=BT_NNL,BTN_NNL=BTN_NNL,BT_BTN=BT_BTN,BT_BTN_RPA=as.data.frame(2**Heatmap_input))
dat<-list(BT_BTN_filter=as.data.frame(BT_BTN_filter))
#BT_BTN_RPA=as.data.frame(2**Heatmap_input)

#the gene of factors were downloaded from the HUMAN PROTEIN ALTLAS
#factors<-c("growth","cytokine","chemokine","prostaglandin","interleukin","integrin","extracellular","metabolism","hormone")
factors<-c("secretome")

#for each factor, differential gene expression using limma
#import secretome gene list 
setwd("C:/Users/37140/Downloads/")
secretome2<-read.csv("C:/Users/37140/Downloads/secretome2.tsv",sep="")
secretome1<-read.csv("C:/Users/37140/Downloads/secretome1.tsv",sep="")
secretome<-data.frame(Gene=c(secretome2$Gene,secretome1$Gene))
#dim(secretome)
#[1] 1708    1
secretome<-distinct(secretome)
#1708 1

for (i in names(dat)){dat[[i]]
group<-as.factor(word(names(dat[[i]]),1,sep="[._]"))
save_dir<-paste0(names(dat[i]),"/")
dir.create(save_dir)
design<-model.matrix(~0+group)
row.names(design)<-colnames(dat[[i]])
colnames(design)<-as.vector(unique(group))
DGElist<-DGEList(counts=dat[[i]],group=group)
DGElist<-calcNormFactors(DGElist)

#expression transformation using voom, another methods is to calculate the logCPM
#logCPM <- cpm(dge, log=TRUE, prior.count=3)
v<-voom(DGElist,design,plot=TRUE,normalize="quantile") 
fit<-lmFit(v,design)
cont.matrix<-makeContrasts(contrasts = paste(colnames(design),collapse = "-"),levels=design)
fit2<-contrasts.fit(fit,cont.matrix)
fit2<-eBayes(fit2)
nrDGE_limma_voom=topTable(fit2,coef=paste(colnames(design),collapse = "-"),n=Inf)
head(nrDGE_limma_voom)
DGE=nrDGE_limma_voom
logFC_cutoff=1  
FDR=0.2
DGE$change = as.factor(ifelse(DGE$P.Value < 0.1 & abs(DGE$logFC) > logFC_cutoff & DGE$adj.P.Val < FDR,
                              ifelse(DGE$logFC > logFC_cutoff ,'UP','DOWN'),'NOT')
)
#library(clusterProfiler)
#library(org.Hs.eg.db)
DGE=DGE[order(DGE$logFC,decreasing = T),]
write.csv(DGE,paste0(save_dir,paste(levels(group),collapse = "-"),".csv"),quote = F) #DGE up and down genes are uncertain secreted protein,and this is a protein

annotation_col = data.frame(TissueType = factor(group)) 
print(annotation_col$TissueType)
rownames(annotation_col) = names(dat[[i]]) #consistent


for (i in factors){
  #factor<-read.csv(paste0(gsub("factor",i,"C:/Users/37140/Downloads/factor"),".","tsv"),sep="")
  DGE_factor<-(DGE[rownames(DGE) %in% intersect(rownames(DGE),factor$Gene),]) #until now, the gene are limited to which produce secreted protein
  head(DGE_factor)
  genes<-row.names(DGE_factor[!(DGE_factor$change=="NOT"),])
  head(annotation_col)
  print(annotation_col)
  #group1<-as.character(annotation_col$TissueType[1])
  #group2<-as.character(annotation_col$TissueType[2])
  annota_colors<-list(TissueType=c("BT"="firebrick","BTN"="pink3","NNL"="white"))
  pheat<-pheatmap(log2(TPM[rownames(TPM) %in% genes,]+1)[,c(rownames(DGElist$samples))],scale = "row", clustering_distance_rows = "correlation",border=FALSE,main = toupper(i),color = colorRampPalette(c("navy", "white", "firebrick3"))(100),annotation_col=annotation_col,annotation_colors=annota_colors)
  save_pheatmap_pdf(pheat, paste0(save_dir,i,".pdf"))
}
}

#the function as saving the pheatmap pdf figures
save_pheatmap_pdf <- function(x, filename,width=10,height=10) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}


annota_colors<-list(TissueType=c("BT"="firebrick","BTN"="blue3"))
unfavo_exp<-log2(TPM[rownames(TPM) %in% factor(unfavo_sig),]+1)
rownames(unfavo_exp)<-factor(rownames(unfavo_exp),levels=unfavo_sig)
pheat1<-pheatmap( unfavo_exp[,c(rownames(DGElist$samples))],scale = "row", clustering_distance_rows = "correlation",border=FALSE,main = toupper(i),annotation_col=annotation_col,annotation_colors=annota_colors, color = colorRampPalette(c("navy", "white", "firebrick3"))(100), fontsize = 15)
save_pheatmap_pdf(pheat1, paste0("unfavo",".pdf"))

#using the normal patients as a control group, compared the relative abundant between tumor and paracancerous tissues 
#keep the genes in secretome,TPM has been filtered 
secretome_exp<-TPM[rownames(TPM) %in% unique(secretome$Gene),] 

#mean_value<-as.data.frame(secretome_exp %>% dplyr::select(starts_with("NNL")) %>% apply(.,1,mean))
secretome_exp<-secretome_exp %>% mutate(mean_NNL = rowMeans(select(., starts_with("NNL")), na.rm = TRUE)) #nrow=874

#mean_value pf tumor and non-tumor samples
secretome_exp<-secretome_exp %>% mutate(mean_BT = rowMeans(select(., starts_with("BT.")), na.rm = TRUE)) 
secretome_exp<-secretome_exp %>% mutate(mean_BTN = rowMeans(select(., starts_with("BTN.")), na.rm = TRUE))

#filtering  the gene expressing lower expression in tumor or non-tumor than normal tissue
secretome_exp<-secretome_exp %>% filter(mean_BT> mean_NNL,mean_BTN> mean_NNL) 

###method3.1 filter out the genes expression lower in tumor or non-tumor tissue
string<-c("mean","NNL")
BT_BTN_filter<-secretome_exp %>% select(!(starts_with(string)))


#filter out the gene without expressing in normal patients 
secretome_exp<-secretome_exp[!(secretome_exp$mean_NNL==0),] #nrow=862,delete about 3 genes

secretome_exp_1 <- secretome_exp %>% mutate(
                                  BT_1vNNL = log((BT.1+0.1)/mean_NNL,2),
                                  BT_2vNNL = log((BT.2+0.1)/mean_NNL,2),
                                  BT_4vNNL = log((BT.4+0.1)/mean_NNL,2),
                                  BT_5vNNL = log((BT.5+0.1)/mean_NNL,2),
                                  BT_6vNNL = log((BT.6+0.1)/mean_NNL,2),
                                  BT_7vNNL = log((BT.7+0.1)/mean_NNL,2),
                                  BT_8vNNL = log((BT.8+0.1)/mean_NNL,2),
                                  BT_9vNNL = log((BT.9+0.1)/mean_NNL,2),
                                  BT_10vNNL = log((BT.10+0.1)/mean_NNL,2),
                                  BT_11vNNL = log((BT.11+0.1)/mean_NNL,2),
                                  BT_13vNNL = log((BT.13+0.1)/mean_NNL,2),
                                  BTN_1vNNL = log((BTN.1+0.1)/mean_NNL,2),
                                  BTN_2vNNL = log((BTN.2+0.1)/mean_NNL,2),
                                  BTN_4vNNL = log((BTN.4+0.1)/mean_NNL,2),
                                  BTN_5vNNL = log((BTN.5+0.1)/mean_NNL,2),
                                  BTN_6vNNL = log((BTN.6+0.1)/mean_NNL,2),
                                  BTN_7vNNL = log((BTN.7+0.1)/mean_NNL,2),
                                  BTN_8vNNL = log((BTN.8+0.1)/mean_NNL,2),
                                  BTN_9vNNL = log((BTN.9+0.1)/mean_NNL,2),
                                  BTN_10vNNL = log((BTN.10+0.1)/mean_NNL,2),
                                  BTN_11vNNL = log((BTN.11+0.1)/mean_NNL,2)
                                  ) 

Heatmap_input <- secretome_exp_1 %>% select(ends_with("vNNL")) 
Heatmap_input<-as.matrix(Heatmap_input)
#use the Heatmap_input to do the methods1 gene differential analysis
gene<-rownames(BT_BTN_RPA[!(BT_BTN_RPA$change=="NOT"),])
Heatmap_input_1<-Heatmap_input[rownames(Heatmap_input) %in% gene,] #nrow 234, The label gene symbolizes the differentially expressed genes in tumor and paracancerous tissues(BT_BTN_RPA)
write.csv(Heatmap_input_1,"./Heatmap_input_1.csv",quote=F)
#Heatmap_input_1<-read.csv("Heatmap_input_1.csv",row.names = 1)

unfavo_sig<-factor(sig[!(sig %in% c("VWF","REN")) ])
unfavo_input<-Heatmap_input_1[rownames(Heatmap_input_1) %in% factor(unfavo_sig), ]
rownames(unfavo_input)<-factor(rownames(unfavo_input),levels=unfavo_sig)


#Derive the colour palette
ColSideColors= c("firebrick","blue3")[as.factor(c(rep("BT",11),rep("BTN",10)))]


#Derive the colour palette
my_palette <- colorRampPalette(c("navy", "white", "firebrick3"))(100)
png("2021_heatmap.png",width=600,height=600,res = 100)
par(mfrow=c(8,8), mar=c(0,1,1,1))
heatmap.2(unfavo_input,main = "",
          notecol="black",
          density.info="density",
          trace="none",
          col=my_palette,
          symbreaks=TRUE,
          dendrogram="both",
          Colv=TRUE,
          Rowv=TRUE,
          scale='none',
          ColSideColors = ColSideColors,
          keysize = 1,
          key.xtickfun = NULL,
          key.ytickfun = NULL,
          key.xlab = "Log2(FC)",
          #xlab = "Subtype",
          ylab = "",
          cexCol = 0.7,
          cexRow =1.2,
          symkey = FALSE,
          #font <- par("font.axis"),
          #scale = c("row"),
          labRow = rownames(unfavo_input)
        )
dev.off()
#Legend_sort <- Heatmap_input[order(Heatmap_input_1$Group),]

#Add legend to the heatmap at correct posotion next to colour key          
legend(y=1.23, x=.008, xpd=TRUE,      # location of the legend on the heatmap plot
       col = col1,  
       lty= 1,             
       lwd = 5,
       cex = .4)

dev.off()
#normlzed<-normlzed[apply(normlzed, 1, function(x) sd(x)!=0.000),]

#differential gene to filter most secretome genes 
BT_BTN_DGE<-read.csv("./BT_BTN/BT-BTN.csv",sep=",",row.names = 1)
BT_NNL_DGE<-read.csv("./BT_NNL/BT-NNL.csv",sep=",",row.names = 1)
BTN_NNL_DGE<-read.csv("./BTN_NNL/BTN-NNL.csv",sep=",",row.names = 1)
BT_BTN_RPA<-read.csv("C:/Users/37140/Documents/BT_BTN_RPA/BT-BTN.csv",sep=",",row.names = 1)

#dim(BT_BTN_RPA[(BT_BTN_RPA$change=="UP"),])
#41,8

###core secretome pca or tsn
DGE_secret<-BT_BTN_DGE[BT_BTN_DGE$X %in% secretome$Gene,]
sub_DGE<-head(DGE_secret[order(abs(DGE_secret$logFC),decreasing = T),],50)
sub_TPM<-as.data.frame(t(TPM[rownames(TPM) %in% sub_DGE$X,] %>% select(.,-starts_with("NNL")))) #remove the normal sample and transpose
sub_TPM$conditions<-c(rep(c("tumor","non_tumor"),c(11,10)))
pca_check2(sub_TPM)

##two methods to divide the tumor and non-tumor sample
pca_check2<-function(sub_TPM){
  library(factoextra)
  library(FactoMineR)
  variableL <- ncol(sub_TPM)
  pca <- prcomp(sub_TPM[,1:(variableL-1)], scale=T) #prcomp: performs a principal components analysis 
  fviz_eig(pca, addlabels = TRUE) 
  fviz_pca_ind(pca, geom.ind="point",col.ind=sub_TPM$conditions, mean.point=F, addEllipses = T, legend.title="Groups",ellipse.type="confidence", ellipse.level=0.95)+ggforce::geom_mark_ellipse(aes(fill = sub_TPM$conditions,color = sub_TPM$conditions)) +theme(legend.position = 'bottom') +coord_equal()
}
ggsave('all_samples_PCA.png')

library(ggplot2)
library(ggfortify)
autoplot(prcomp(sub_TPM[,1:(ncol(sub_TPM)-1)] ), data=sub_TPM,colour = 'conditions')+theme_bw() 


#KEGG analysis
library(clusterProfiler)
Genelist<-BT_BTN_DGE
entrezIDs <- bitr(geneID = Genelist$X,fromType = "SYMBOL",toType = "ENTREZID",
                  OrgDb ="org.Hs.eg.db" ,drop = T)

KErt <- Genelist[Genelist$X %in% entrezIDs$SYMBOL,]$X
KErt <- entrezIDs[entrezIDs$SYMBOL%in%KErt,]$ENTREZID

KEGGa <- enrichKEGG(KErt, organism = 'hsa', keyType = 'kegg', 
                    pvalueCutoff = 0.05,
                    pAdjustMethod = 'BH', 
                    minGSSize =1,maxGSSize = 10000,qvalueCutoff = 0.1,
                    use_internal_data = F)
barplot(KEGGa,drop=T,showCategory = 30)
dotplot(KEGGa,showCategory=30)




